### PR TITLE
colorization util function to help activation map uploading

### PR DESCRIPTION
**Pull request by y27choi**
[activation map example](https://trunk.kolena.io/yoohee/studio?workflow=Person+Detection&testSuiteId=1314&testCaseId=20134&models=N4IglgJiBcBs8BoQGMD2A7AZmA5gVwCcBDAFzAxlEhgE4aAmAX0aA&overlays=N4Igtg9gJgpgNgZxALmAXwDQgIYDtcQAu2hAlhLkqiITAoQPoLZgAOcMDYMhATqQGMqoAGIA5FKAAWRANYwAnihABGEFgBupBKQBGHFHwCuMTCBEAFSSBmF5S5CABM6kFp36Yh3ibMAVK2pbe2UAZld3PQNkY1MzUlwAMxheGFwBL2pdXQgADzprYMVlABYI7SjMxOxEUywpGBIwbFZhGzlixwBWco9o2LQzAHNeCCNcKAZjQilrbLyCoI6HEAA2XsrvXyxE0iGjVLailYB2Dc8tupB7VggEwiPl5QAOc-6fOKxaeiYWdkz0F86IxmGwOFweNgoCRsJJBmggA&testSampleId=16750231) colorized using the new util function

### Linked issue(s):
Fixes KOL-2658

### What change does this PR introduce and why?
Added an utility function to help users upload activation map easily with the default colormap and alpha adjustment

### Please check if the PR fulfills these requirements

- [ ] Include reference to internal ticket and/or GitHub issue "Fixes #NNNN" (if applicable)
- [ ] Relevant tests for the changes have been added
- [ ] Relevant docs have been added / updated

### CODE REVIEW

File: kolena/workflow/visualization/_activation_map.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        return self._scale_to_colormap(intensity - MAX_UINT8 / 4)
+
+    def green(self, intensity: np.uint8) -> float:
+        return self._scale_to_colormap(intensity)
+
+    def blue(self, intensity: np.uint8) -> float:
+        return self._scale_to_colormap(intensity + MAX_UINT8 / 4)
+
+
+def colorize_activation_map(activation_map: np.ndarray, fade_low_activation: bool = True) -> np.ndarray:
```
Comment (y27choi) [inline kolena/workflow/visualization/_activation_map.py]: Note to self: add a description

File: kolena/workflow/visualization/_activation_map.py [hunk 2]
```diff
@@ -70,6 +70,17 @@ def blue(self, intensity: np.uint8) -> float:
 
 
 def colorize_activation_map(activation_map: np.ndarray, fade_low_activation: bool = True) -> np.ndarray:
+    """
+    Applies the [MATLAB "Jet" colormap](http://blogs.mathworks.com/images/loren/73/colormapManip_14.png) to the
+    activation map.
+
+    :param np.ndarray activation_map: A 2D numpy array, shaped (h, w) or (h, w, 1), of the activation map in `np.uint8`
```
Comment (gordonhart) [inline kolena/workflow/visualization/_activation_map.py]: The additional `np.ndarray` isn't necessary — type is inferred from type hints. Just needs `:param activation_map:`

File: kolena/workflow/visualization/_activation_map.py [hunk 3]
```diff
@@ -70,6 +70,17 @@ def blue(self, intensity: np.uint8) -> float:
...
+        or `float` ranging [0, 1].
```
Comment (gordonhart) [inline kolena/workflow/visualization/_activation_map.py:123]: Does `[0, 1]` only apply to float type inputs? Are floats of any precision accepted?

File: kolena/workflow/visualization/_activation_map.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+            intensity = pixel[0]
+            activation_map[i][j] = np.array(
+                [
+                    colormap.blue(intensity),
+                    colormap.green(intensity),
+                    colormap.red(intensity),
+                    pixel[3] if fade_low_activation else MAX_UINT8,
+                ],
+                dtype=np.uint8,
+            )
```
Comment (gordonhart) [inline kolena/workflow/visualization/_activation_map.py]: Is there any way for us to apply this as a vectorized function? I would expect nested `for` loops to be quite slow.

File: kolena/workflow/visualization/_activation_map.py (hunk 3 above)
Comment (y27choi) [reply kolena/workflow/visualization/_activation_map.py:123]: Yes and yes

File: kolena/workflow/visualization/_activation_map.py [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from abc import ABC
+from abc import abstractmethod
+
+import numpy as np
+
+from kolena.errors import InputValidationError
+
+MAX_UINT8: np.uint8 = 255
```
Comment (gordonhart) [inline kolena/workflow/visualization/_activation_map.py]: Avoid declaring superfluous const via `np.iinfo(np.uint8).max`?

File: kolena/workflow/visualization/_activation_map.py [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+# limitations under the License.
+from abc import ABC
+from abc import abstractmethod
+
+import numpy as np
+
+from kolena.errors import InputValidationError
+
+
+class ColorMap(ABC):
```
Comment (gordonhart) [inline kolena/workflow/visualization/_activation_map.py]: Nit: we should use either "Color Map" (space) or "Colormap" (no space) consistently across names and documentation — there's a bit of both currently

File: kolena/workflow/visualization/_activation_map.py [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+        return self._scale_to_colormap(intensity - np.iinfo(np.uint8).max / 4)
+
+    def green(self, intensity: np.uint8) -> float:
+        return self._scale_to_colormap(intensity)
+
+    def blue(self, intensity: np.uint8) -> float:
+        return self._scale_to_colormap(intensity + np.iinfo(np.uint8).max / 4)
+
+
+def colorize_activation_map(activation_map: np.ndarray, fade_low_activation: bool = True) -> np.ndarray:
```
Comment (gordonhart) [inline kolena/workflow/visualization/_activation_map.py]: Not critical for now but it seems to me an improved interface would be to have `colormap` as a kwarg (w/ default) and bake the `fade_low_activation` into the colormap. That allows users to plug in their own colors, if desired, and keeps this function simple.

File: kolena/workflow/visualization/_activation_map.py (hunk 6 above)
Comment (y27choi) [reply kolena/workflow/visualization/_activation_map.py]: Thanks for catching this. I will stick to "Colormap"

File: kolena/workflow/visualization/_activation_map.py (hunk 7 above)
Comment (y27choi) [reply kolena/workflow/visualization/_activation_map.py]: Ah I like that. Thanks for the suggestion. Will make that change.

File: kolena/workflow/visualization/_activation_map.py [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+    if fade_low_activation:
+        # apply non-linear scaling alpha channel [0, 255]
+        alpha = max_uint8 / (1 + np.exp(((max_uint8 / 2) - activation_map[:, :, 0]) / 8))
+        activation_map[:, :, 3] = alpha
+    else:
+        activation_map[:, :, 3] = max_uint8
+
+    activation_map[:, :, 0] = vblue(activation_map[:, :, 0])
+    activation_map[:, :, 1] = vgreen(activation_map[:, :, 1])
+    activation_map[:, :, 2] = vred(activation_map[:, :, 2])
```
Comment (gordonhart) [inline kolena/workflow/visualization/_activation_map.py]: Would it be simpler to have the `Colormap` class expose a single function that returns the RGB pixel as a single piece? Would simplify the vectorization and application to something like this:

```py
vcolor = np.vectorize(ColormapJet().colorize)
activation_map[: ,:] = vcolor(activation_map[:, :])
```

**Review by gordonhart — APPROVED** [review]
A few minor comments, otherwise looks good :+1:

**Comment by y27choi** [comment]
Addressed all of your comments, @gordonhart.  Do you mind taking another look at it before I merge it?

**Review by gordonhart — APPROVED** [review]

**Pull request merged**